HPCC-34746 Allow cost definitions for agent, manager and workers in Thor's yaml

shamser opened the pull request:
…

This commit introduces support for defining costs for Thor cluster components:
- Added cost configuration support for Thor agent, manager, and worker processes
- Modified helm templates to handle cost definitions in thor.yaml
- Updated values.yaml and schema.json to support new cost configuration options
- Enhanced workunit and agent code to utilize cost information
- Added utility functions in jlib to support cost calculations

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

shamser committed:
HPCC-34746 Allow cost definitions for agent, manager and workers in Thor's yaml

- Added cost definitions for agent, manager and workers in Thor's yaml
- The new component level cost definition overrides existing thor cost definition
- Warn if the thor level cost definition is not consistent with the component level cost definition
- Warn if thor level cost definition is being used
- Updated the values.yaml with the new cost definitions (commented out)

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexisrisk.com>

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -14220,8 +14220,16 @@ extern WORKUNIT_API void addTimeStamp(IWorkUnit * wu, unsigned wfid, const char
+        double costCpuHour = getComponentConfigSP()->getPropReal(subComponentSection.str(), -1.0);
+        if (costCpuHour >= 0.0)
+            return costCpuHour;
+        // don't have sub component cost, drop-through to use component/global cost values
```
I think this should be an error, and the templates should ensure that the sub section is generated.
i.e. after this PR, Thor should always generate a "eclagent", "manager", and, "worker" sections.
The legacy support for the old "cost" can be handled in the templates, allowing this code to be simplified/standardized on the new schema.

NB: the chart version must match the image version.

shamser replied:
Thor will have the cost section in the sub-section.   It is not necessary the case for other components such as eclagent and eclcc.  I've created Jira to make this change: https://hpccsystems.atlassian.net/browse/HPCC-35054

jakesmith commented on helm/hpcc/values.schema.json:
```diff
@@ -929,6 +929,16 @@
       "additionalProperties": false
     },
+    "thorSubComponent": {
+      "description": "Thor sub-component with resource and cost configuration",
```
currently only cost configuration, not resources? But (see other comment) it would be good to standarize on this schema, i.e. move "resources" under the subcomponent (but in separate JIRA).

shamser replied:
To keep the PR small and simply, I have limited the changes in the PR to adding the cost definitions only. The other changes as suggested about is planned in a separate PR (I'll be adding additional items under the parent ticket: https://hpccsystems.atlassian.net/browse/HPCC-34436)

jakesmith reviewed (changes requested):
@shamser - please see comments.

shamser commented on helm/hpcc/values.yaml:
```diff
@@ -895,6 +895,15 @@ thor:
+  #    perCpu: 0.113
+  #eclAgent:
+  #  cost:
+  #    perCpu: 0.053
```
It seems to be working for me.  I'll run some more tests.

shamser committed:
Changes following review

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexisrisk.com>

jakesmith reviewed (changes requested):
@shamser - looks good in general. Please see comments.

A couple of comments re. tests, that related to new checks in _warnings.tpl.
I think it should be invalid for both "cost" and one of the new subsections with cost within them to be present.
Either it's an old legacy values yaml with a singleton "cost", or they have migrated to new schema with separate subcomponent cost sections - if they create one with a mixture we can reject it with a at deployment error.

shamser commented on helm/hpcc/templates/thor.yaml:
```diff
@@ -21,18 +21,49 @@
-{{- $eclAgentScope := dict "name" .eclAgentName "type" $eclAgentType "useChildProcesses" .eclAgentUseChildProcesses "replicas" .eclAgentReplicas "maxActive" .me.maxJobs | merge (pick .me "auxQueues" "expert" "keepJobs" "logging" "tracing" "cost" "schedulingTimeoutSecs") ((hasKey .me "holdAgent") | ternary (dict "hold" .me.holdAgent) dict) }}
-{{- $thorAgentScope := dict "name" .thorAgentName "targetName" .me.name "replicas" .thorAgentReplicas "maxActive" .me.maxGraphs | merge (pick .me "expert" "keepJobs" "logging" "tracing" "schedulingTimeoutSecs") ((hasKey .me "holdThorAgent") | ternary (dict "hold" .me.holdThorAgent) dict) }}
+{{- $eclAgentCostConfig := include "generateThorComponentCostConfig" (dict "root" .root "me" .me "subComponent" "eclagent") | fromYaml -}}
+{{- $eclAgentCost := $eclAgentCostConfig.cost | default dict -}}
```
If there is no costs provided, then an empty dict will be returned.  This mean that although $eclAgentCostConfig is a dict,  $eclAgentCostConfig.cost doesn't necessarily exist.

shamser commented on helm/hpcc/templates/thor.yaml:
```diff
@@ -21,18 +21,49 @@
+{{- $managerCostConfig := include "generateThorComponentCostConfig" (dict "root" .root "me" .me "subComponent" "manager") | fromYaml -}}
+{{- $managerCost := $managerCostConfig.cost | default dict -}}
+{{- $workerCostConfig := include "generateThorComponentCostConfig" (dict "root" .root "me" .me "subComponent" "worker") | fromYaml -}}
+{{- $workerCost := $workerCostConfig.cost | default dict -}}
```
generateThorComponentCostConfig does return a dict but it doesn't necessarily return .dict.

shamser committed:
Changes following review

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexisrisk.com>

jakesmith commented on helm/hpcc/templates/thor.yaml:
```diff
@@ -76,7 +107,7 @@ data:
       type: thor
     global:
-{{ include "hpcc.generateGlobalConfigMap" .root| indent 6 }}
+{{ include "hpcc.generateGlobalConfigMap" (dict "root" .root "excludeCost" true) | indent 6 }}
```
what would happen if cost wasn't excluded in global config here?

jakesmith reviewed (commented):
@shamser - see a couple of new comments.

But I have not studied (or marked resolved) changes (if there are some) related to my previous comments, because I don't know whether you've looked at them/addressed them.
Please comment on each of my comments, or acknowledge them (e.g. thumbs up = agree/to do). 
Same is true of the copilot comments.

Also, please fill in the JIRA with a description detailing the goal/approach etc.

shamser commented on helm/hpcc/templates/thor.yaml:
```diff
@@ -21,18 +21,49 @@
 {{- $eclAgentResources := .me.eclAgentResources | default dict -}}
-{{- $hthorScope := dict "name" $hthorName "jobMemorySectionName" "eclAgentMemory" | merge (pick .me "expert" "logging" "maxGraphStartupTime" "tracing" "cost" "schedulingTimeoutSecs") | merge (dict "resources" (deepCopy $eclAgentResources))  ((hasKey .me "holdWorkflow") | ternary (dict "hold" .me.holdWorkflow) dict)}}
-{{- $thorScope := omit .me "eclagent" "thoragent" "hthor" "env" "eclAgentResources" "eclAgentUseChildProcesses" "eclAgentReplicas" "thorAgentReplicas" "eclAgentType" }}
+{{- $hthorScope := dict "name" $hthorName "jobMemorySectionName" "eclAgentMemory" | merge (pick .me "expert" "logging" "maxGraphStartupTime" "tracing" "cost" "schedulingTimeoutSecs") | merge (dict "resources" (deepCopy $eclAgentResources))  ((hasKey .me "holdWorkflow") | ternary (dict "hold" .me.holdWorkflow) dict )}}
```
Done.

shamser commented on helm/hpcc/values.schema.json:
```diff
@@ -2883,8 +2893,19 @@
-          "$ref" : "#/definitions/componentCost"
+          "$ref" : "#/definitions/componentCost",
+          "description": "deprecated cost section, use manager, worker, and eclagent sections",
+          "type": "object"
```
Done.

shamser commented on testing/helm/tests/thor-subcomponent-inherit.yaml:
```diff
@@ -0,0 +1,28 @@
+thor:
+- name: thor
+  cost:   # Only worker overrides; manager and eclagent should inherit this value
+    perCpu: 1.80
```
I agree. I'll remove support for depreciated thor-level cost.

shamser committed:
Changes following review

1. Schema Changes (values.schema.json)
Removed the deprecated cost property from the thor component definition
2. Template Changes (thor.yaml)
Updated generateThorComponentCostConfig template to remove fallback to thor.cost
Cost inheritance now follows a simpler pattern:
Subcomponent explicit cost (if defined)
Falls back to global.cost.perCpu (if defined)
Removed "cost" from the omit list in $thorScope construction since it's no longer a valid thor-level property
3. Warning Template Changes (_warnings.tpl)
Removed warning logic for thor-level vs subcomponent-level cost conflicts
Removed deprecation warning for using thor.cost
Kept the warning for inconsistent subcomponent costs (when some subcomponents have cost defined but others don't)
4. Add cost section to thor's hthor section
5. Test File Updates
thor-cost-explicit.yaml (renamed from thor-cost-conflict.yaml):

Removed deprecated thor-level cost
Now tests explicit cost definitions for all three subcomponents
Renamed to better reflect its purpose
thor-subcomponent-inherit.yaml:

Updated to demonstrate inheritance from global.cost instead of thor.cost
Worker has explicit cost; manager and eclagent inherit from global
Updated comments to reflect new behavior

REDBOOK NOTE:
- Breaking change: thor.cost has been depreciated and is no longer supported. Users must migrate to subcomponent-level cost definitions.

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexisrisk.com>

jakesmith reviewed (commented):
@shamser - 1 comment, looks good I think.

jakesmith commented:
@ghalliday - please sanity check.

shamser committed:
Retain depreciated thor-level cost in schema

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexisrisk.com>
